### ISSUE TITLE
Enable the use of multiple arcs for staging

### ISSUE THREAD

**Issue by SeanMatthewNolan** [Rapid-Design-of-Systems-Laboratory/beluga#123]
I need this for certain problems, and it will be useful for problems like rocket staging. I hope to be able to add it in such a way that only optional arguments are needed, so that all previous single arc problems behave the same. Theoretically, changes should nearly exclusively occur in optimlib, so nothing in the actually solving should change.

**Comment by msparapa** [comment]
Right. I think the bvp solver should solve all problems pretty much exactly the way it currently does. I can't really think of any additional changes it would need. The extra arcs should just be more rows in the EOM function. So a single stage to orbit vehicle might have 4 EOMs, but a two stage to orbit vehicle may have 8 EOMs all propagated together. So it's not truly multiple shooting in the computational sense, but adding arcs for multiple shooting is already available on the bvp solver side of things.

I added a time optimal launch problem as a project on github. Ideally we would have a staged rocket as an example problem, but this is going to take some significant work. I've already added a few issues that need to be resolved first. Add more as you see fit.

**Comment by SeanMatthewNolan** [comment]
I forgot that I already submitted this as #51. I apologize for that.

**Comment by msparapa** [comment]
Changed the names so they're different now.

**Comment by msparapa** [comment]
I'm working on this right now using Harish's RASHS method. I've figured out how to model the switch in dynamics, i.e. how to have a mass flow rate drop from -1 to -0.25 to represent a second stage burning less fuel. But how do I model the mass lost over the "discontinuity"? So say mass has gone from 1 to 0.6 during stage 1, when stage 2 starts, the mass will be closer to 0.55 because of the dropped engines from stage 1. I don't think Harish ever specifically addressed this in his paper, but he mentioned its possible. Do you know how to model this?

**Comment by msparapa** [comment]
Actually I think I just figured it out. In addition to stage 1 and stage 2, should I add a third "stage 1.5" with a high mass flow rate but no thrust? This is the only way I can think of to handle this.

**Comment by SeanMatthewNolan** [comment]
I actually solved a problem like that by using a constant term that is switched off. For example if f(x) is the switching function, I had mass = f(x)*staging_mass + variable_fuel_mass.

**Comment by msparapa** [comment]
Huh neat. It looks like with multiple solutions to this problem then I'm not going to tackle that particular issue upfront. Rather I'll leave it open to the user on how they want to implement the discontinuity, for now at least.

**Comment by msparapa** [comment]
A few things I need some input on:

1. I think staging will be best implemented under the quantity method instead of directly in the state rates. We discussed before having something like:

```
ocp.state('x', ['xdot1', 'xdot2'], 'm')
```

but after learning Harish's method and thinking about this for a bit:

```
ocp.state('x', 'xdot', 'm')
ocp.quantity('xdot', ['xdot1', 'xdot2'], [condition1, condition2], tolerance='s')
```

would be better. It's a slightly more general switching structure than switching the state rates themselves since you can switch the quantities contained within. Also...

2. Changing the name of `quantity` to `switch`. This better implies what the unique purpose of quantities are. Previous functionality like we had before:

```
ocp.switch('D', '0.5*rho*v^2*Cd*Aref')
```

will be unchanged. Maybe I'll make a special `Always` condition as a constant in the utils folder or something similar.

3. Changing the name of `activator` to `tolerance` in path constraints and switch (like above). In switch, I'll internally use `1/s` instead of `s` directly, that way as `s -> 0`, the discontinuity between stages becomes sharper. I had the idea to call them activators before because they "activate" constraints, but that's not really true. If anything, the constraints start off over-activated. Calling them a tolerance isn't exactly correct either, but at least it implies how they are used where smaller number = higher quality.
